pinot-core/src/main/java/org/apache/pinot/core/io/util/VarLengthBytesValueReaderWriter.java

```diff
@@ -0,0 +1,166 @@
+  }
+
+  private void writeHeader() {
+    for (int offset = 0; offset < MAGIC_HEADER_LENGTH; offset++) {
+      _dataBuffer.putByte(offset, MAGIC_HEADER[offset]);
+    }
+  }
+
+  public void init(byte[][] byteArrays) {
+    this._numElements = byteArrays.length;
+
+    writeHeader();
+
+    // Add the number of elements as the first field in the buffer.
+    _dataBuffer.putInt(MAGIC_HEADER_LENGTH, byteArrays.length);
+
+    // Then write the offset of each of the byte array in the data buffer.
+    int nextOffset = INDEX_ARRAY_START_OFFSET;
+    int nextArrayStartOffset = INDEX_ARRAY_START_OFFSET + Integer.BYTES * byteArrays.length;
+    for (byte[] array: byteArrays) {
+      _dataBuffer.putInt(nextOffset, nextArrayStartOffset);
+      nextOffset += Integer.BYTES;
+      nextArrayStartOffset += array.length;
+    }
+
+    // Finally write the byte arrays.
+    nextArrayStartOffset = INDEX_ARRAY_START_OFFSET + Integer.BYTES * byteArrays.length;
+    for (byte[] array: byteArrays) {
+      _dataBuffer.readFrom(nextArrayStartOffset, array);
+      nextArrayStartOffset += array.length;
+    }
+  }
+
+  int getNumElements() {
+    // Lazily initialize the numElements.
+    if (_numElements == 0) {
+      _numElements = _dataBuffer.getInt(MAGIC_HEADER_LENGTH);
+      Preconditions.checkArgument(_numElements > 0);
+    }
+    return this._numElements;
+  }
+
+  @Override
+  public int getInt(int index) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public long getLong(int index) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public float getFloat(int index) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public double getDouble(int index) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public String getUnpaddedString(int index, int numBytesPerValue, byte paddingByte, byte[] buffer) {
+    return StringUtil.decodeUtf8(getBytes(index, numBytesPerValue, buffer));
+  }
+
+  @Override
+  public String getPaddedString(int index, int numBytesPerValue, byte[] buffer) {
+    return StringUtil.decodeUtf8(getBytes(index, numBytesPerValue, buffer));
+  }
+
+  @Override
+  public byte[] getBytes(int index, int numBytesPerValue, byte[] buffer) {
+    // Read the offset of the byte array first and then read the actual byte array.
+    int offset = _dataBuffer.getInt(INDEX_ARRAY_START_OFFSET + Integer.BYTES * index);
+
+    // To get the length of the byte array, we use the next byte array offset. However,
+    // for last element there is no next byte array so we use buffer size.
+    int length;
+    if (index == getNumElements() - 1) {
+      // This is the last byte array in the buffer so use buffer size to get the length.
+      length = (int) (_dataBuffer.size() - offset);
+    }
+    else {
+      length = _dataBuffer.getInt(INDEX_ARRAY_START_OFFSET + Integer.BYTES * (index + 1)) - offset;
+    }
+
+    byte[] b = buffer == null ? new byte[length] :
```

Comment:
The callers of this method might pass a null byte buffer too and we create a new buffer here in that case. The caller knows the buffer length in advance for fixed length dictionary hence it can create and pass the buffer but it's not possible for var length dict, hence we create the buffer here.

I'll simplify the code for readability.

Reply:
This is done.